File: claude/web/src/lib/arcade/registry.ts
/**
 * Env registry: map a stable `env_id` to its env class.
 *
 * Mirrors `harness/steambench_harness/registry.py`. The verifier instantiates
 * an env by id so a submitted run can be replayed without trusting the client
 * about which env it ran on.
 */

import type { Env } from "./base";
import type { EnvSpec } from "./types";
import { Catcher } from "./catcher";
import { Connect4 } from "./connect4";
import { Dodger } from "./dodger";
import { Forager } from "./forager";
import { Flappy } from "./flappy";
import { Game2048 } from "./game2048";
import { Minesweeper } from "./minesweeper";
import { Phantom } from "./phantom";
import { Rally } from "./rally";
import { Snake } from "./snake";
import { Sokoban } from "./sokoban";
import { Storm } from "./storm";
import { Tetris } from "./tetris";
import { Turret } from "./turret";
import { Volley } from "./volley";

type EnvCtor = new () => Env;

const REGISTRY: Record<string, EnvCtor> = {
  [Catcher.env_id]: Catcher,
  [Connect4.env_id]: Connect4,
  [Dodger.env_id]: Dodger,
  [Forager.env_id]: Forager,
  [Flappy.env_id]: Flappy,
  [Game2048.env_id]: Game2048,
  [Minesweeper.env_id]: Minesweeper,
  [Phantom.env_id]: Phantom,
  [Rally.env_id]: Rally,
  [Snake.env_id]: Snake,
  [Sokoban.env_id]: Sokoban,
  [Storm.env_id]: Storm,
  [Tetris.env_id]: Tetris,
  [Turret.env_id]: Turret,
  [Volley.env_id]: Volley,
};

/** Instantiate an env by id. */
export function make(envId: string): Env {
  const ctor = REGISTRY[envId];
  if (!ctor) {
    const known = Object.keys(REGISTRY).sort().join(", ");
    throw new Error(`unknown env_id ${JSON.stringify(envId)}; registered: ${known}`);
  }
  return new ctor();
}

/** Sorted list of registered env ids (mirrors `all_env_ids`). */
export const ENV_IDS: string[] = Object.keys(REGISTRY).sort();

/** Spec dicts for every registered env (drives the catalog + docs). */
export function allSpecs(): EnvSpec[] {
  return ENV_IDS.map((id) => make(id).spec());
}

File: claude/harness/steambench_harness/registry.py
"""Env registry: map a stable ``env_id`` to its env class.

Both the agent runner and the server-side verifier use this to instantiate an
env by id, so a submitted run can be replayed without trusting the client about
which env it ran on.
"""

from __future__ import annotations

from typing import Type

from .protocol import Env

_REGISTRY: dict[str, Type[Env]] = {}


def register(env_cls: Type[Env]) -> Type[Env]:
    """Class decorator that registers an env under its ``env_id``."""
    env_id = env_cls.env_id
    if env_id in _REGISTRY and _REGISTRY[env_id] is not env_cls:
        raise ValueError(f"env_id {env_id!r} already registered")
    _REGISTRY[env_id] = env_cls
    return env_cls


def make(env_id: str) -> Env:
    """Instantiate an env by id."""
    if env_id not in _REGISTRY:
        raise KeyError(f"unknown env_id {env_id!r}; registered: {sorted(_REGISTRY)}")
    return _REGISTRY[env_id]()


def all_env_ids() -> list[str]:
    return sorted(_REGISTRY)


def all_specs() -> list[dict]:
    """Spec dicts for every registered env (drives the catalog + docs)."""
    return [make(env_id).spec() for env_id in all_env_ids()]
